core/src/databases/database.rs

```diff
@@ -52,51 +53,249 @@ impl Database {
-
-                Ok(DatabaseSchema(
-                    rows.into_par_iter()
-                        .map(|(table, rows)| {
-                            Ok((
-                                table.table_id().to_string(),
-                                DatabaseSchemaTable::new(table, TableSchema::from_rows(&rows)?),
-                            ))
-                        })
-                        .collect::<Result<HashMap<_, _>>>()?,
-                ))
+                        .map(|(table, rows)| (table.table_id().to_string(), rows))
+                        .collect::<HashMap<_, _>>()),
+                    _ => Err(anyhow!("Error retrieving rows from database.")),
+                }?;
+                utils::done(&format!(
+                    "DSSTRUCTSTAT Finished retrieving rows: duration={}ms",
+                    utils::now() - time_get_rows_start
+                ));
+
+                let generate_create_table_sql_start = utils::now();
+                let create_tables_sql: String = schema
+                    .iter()
+                    .filter(|(_, table)| !table.schema.is_empty())
+                    .map(|(table_name, table)| {
+                        table
+                            .schema
+                            .get_create_table_sql_string(table_name.as_str())
+                    })
+                    .collect::<Vec<_>>()
+                    .join("\n");
+                utils::done(&format!(
+                    "DSSTRUCTSTAT Finished generating create table SQL: duration={}ms",
+                    utils::now() - generate_create_table_sql_start
+                ));
+
+                let conn = rusqlite::Connection::open_in_memory()?;
+
+                let create_tables_execute_start = utils::now();
+                conn.execute_batch(&create_tables_sql)?;
+                utils::done(&format!(
+                    "DSSTRUCTSTAT Finished creating tables: duration={}ms",
+                    utils::now() - create_tables_execute_start
+                ));
+
+                let insert_execute_start = utils::now();
+                rows_by_table
+                    .iter()
+                    .filter(|(_, rows)| !rows.is_empty())
+                    .map(|(table_name, rows)| {
+                        let table_schema = match schema.get(table_name) {
+                            Some(s) => Ok(s),
+                            None => Err(anyhow!("No schema found for table {}", table_name)),
+                        }?;
+
+                        let (sql, field_names) = table_schema.schema.get_insert_sql(table_name);
+                        let mut stmt = conn.prepare(&sql)?;
+
+                        rows.par_iter()
+                            .map(|r| table_schema.schema.get_insert_params(&field_names, r))
+                            .collect::<Result<Vec<_>>>()?
+                            .iter()
+                            .map(|values| match stmt.execute(params_from_iter(values)) {
```

Comment:
I think you want to run that at the end out of the par_iter. Requires benchmarking.

I suspect this inherently prevents parallelization because the pool will get filled with rows waiting for execution on the lock that protects that stmt.execute. 

I would instead create all params in parallalel and then run the stmt.execute sequentially which is locked anyway.

Can we benchmark the two approach to convince ourselves of the best approach?

Reply:
This is worth the time since this code is likely to be final

Reply:
Already tried that, params_from_iter complains when it is in the par_iter. I can try again again but I think it's not possible because it needs to own the value (so we'd have to copy which I believe is worse ?).

creating and executing the smt cannot be done in the par_iter because conn cannot go through threads.

I am happy to benchmark, but I don't see how copying every single row's data can be faster than calling params from iter sequentially ?

Reply:
Yet it's possible: https://github.com/dust-tt/dust/pull/2222/files#diff-bee624fbe402bae0a0c5cbaa790eda4654833e56e5482f62da6ff74d7fb253abR105

Reply:
No copying involved

Reply:
to clarify:
<img width="1348" alt="Screenshot 2023-11-16 at 13 08 21" src="https://github.com/dust-tt/dust/assets/14199823/ded119c0-d79f-483c-990f-a1900a1b5be2">

Reply:
so I'd have to `clone()` the params (sequentially).

Reply:
It works with `into_iter` instead of `iter` 🤔

Reply:
is the former copying the data ? Or merely saying that from this point only the new interator can reference it ?

Reply:
Ok that actually makes sense